tidy3d/components/parameter_perturbation.py

```diff
@@ -0,0 +1,722 @@
+
+        temperature_range = (
+            float(np.min(perturbation_values.coords["T"])),
+            float(np.max(perturbation_values.coords["T"])),
+        )
+
+        values.update({"temperature_range": temperature_range})
+
+        return values
+
+    @ensure_temp_in_range
+    def sample(
+        self, temperature: Union[ArrayLike[float], SpatialDataArray]
+    ) -> Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]:
+        """Sample perturbation.
+
+        Parameters
+        ----------
+        temperature : Union[ArrayLike[float], SpatialDataArray]
+            Temperature sample point(s).
+
+        Returns
+        -------
+        Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]
+            Sampled perturbation value(s).
+        """
+
+        t_range = self.temperature_range
+        temperature_clip = np.clip(temperature, t_range[0], t_range[1])
+        return self.perturbation_values.interp(T=temperature_clip).drop_vars("T")
+
+    @cached_property
+    def is_complex(self) -> bool:
+        """Whether perturbation is complex valued."""
+        return np.iscomplexobj(self.perturbation_values.data)
+
+
+HeatPerturbationType = Union[LinearHeatPerturbation, CustomHeatPerturbation]
+
+
+""" Elementary charge perturbation classes """
+
+
+def ensure_charge_in_range(
+        sample: Callable[
+        [Union[ArrayLike[float], SpatialDataArray], Union[ArrayLike[float], SpatialDataArray]],
+        Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray],
+    ]
+) -> Callable[
+        [Union[ArrayLike[float], SpatialDataArray], Union[ArrayLike[float], SpatialDataArray]],
+        Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray],
+]:
+    """Decorate ``sample`` to log warning if charge supplied is out of bounds."""
+
+    @functools.wraps(sample)
+    def _sample(
+        self,
+        electron_density: Union[ArrayLike[float], SpatialDataArray],
+        hole_density: Union[ArrayLike[float], SpatialDataArray]
+    ) -> Union[ArrayLike[float], ArrayLike[Complex], SpatialDataArray]:
+        """New sample function."""
+
+        if self.electron_range:
+            e_min, e_max = self.electron_range
+
+            # if temperature range not present just return original function
+            if np.any(electron_density < e_min) or np.any(electron_density > e_max):
+                log.warning(
+                    "Electron density values passed to 'ChargePerturbation.sample()'"
+                    f"is outside of 'ChargePerturbation.electron_range' = {self.electron_range}"
+                )
+
+        if self.hole_range:
+            h_min, h_max = self.hole_range
+
+            # if temperature range not present just return original function
+            if np.any(hole_density < h_min) or np.any(hole_density > h_max):
+                log.warning(
+                    "Hole density values passed to 'ChargePerturbation.sample()'"
```

Comment:
noticed some inconsistency regarding periods at end of messages and docstring descriptions. let's try to add them to all.